Kernel/Devices/USB/UHCIController.cpp

```diff
@@ -312,62 +314,188 @@ void UHCIController::start()
     dbgln("UHCI: Started");
 }
 
-struct setup_packet {
-    u8 bmRequestType;
-    u8 bRequest;
-    u16 wValue;
-    u16 wIndex;
-    u16 wLength;
-};
+TransferDescriptor* UHCIController::create_transfer_descriptor(USBPipe& pipe, PacketID direction, size_t data_len)
+{
+    TransferDescriptor* td = allocate_transfer_descriptor();
+    if (td == nullptr) {
+        return nullptr;
+    }
+
+    u16 max_len = (data_len > 0) ? (data_len - 1) : 0x7ff;
+
+    // Set the token dword
+    td->set_token((max_len << TD_TOKEN_MAXLEN_SHIFT) | ((pipe.data_toggle() ? 1 : 0) << TD_TOKEN_DATA_TOGGLE_SHIFT) | (pipe.endpoint_address() << TD_TOKEN_ENDPOINT_SHIFT) | (pipe.device_address() << TD_TOKEN_DEVICE_ADDR_SHIFT) | (static_cast<u8>(direction)));
+
+    if (pipe.data_toggle())
+        pipe.set_toggle(false);
+    else
+        pipe.set_toggle(true);
+
+    // Set some bits in the control/status register
+    if (pipe.type() == USBPipe::Type::Isochronous) {
+        td->set_isochronous();
+    } else {
+        if (direction == PacketID::IN) {
+            td->set_short_packet_detect();
+        }
+    }
+
+    // Set lowsped bit if the device connected to port is a Lowspeed device (proably unlikely...)
+    if (pipe.device_speed() == USBPipe::DeviceSpeed::LowSpeed) {
+        td->set_lowspeed();
+    }
+
+    td->set_active();
+    td->set_error_retry_counter(3);
```

Comment:
Ah, my bad. I probably should have documented this a bit better instead of making it so magical. This value is specifically a 2bit down counter that's decremented when specific errors occurs. If it hits zero, =the controller generates "stalls" the TD (which I don't believe we handle just yet). Relevant datasheet snip attached.

![image](https://user-images.githubusercontent.com/4768900/120101646-59224d80-c18a-11eb-8f7b-047ab74df227.png)